src/zope/server/dualmodechannel.py

```diff
@@ -208,4 +210,11 @@ def close(self):
         # descriptor.
         assert self.async_mode
         self.connected = False
-        asyncore.dispatcher.close(self)
+        try:
+            asyncore.dispatcher.close(self)
+        except AttributeError: # pragma: no cover (tox coverage environment is python 3)
+            # On Python 2.7, this is not idempotent. If we were
+            # already closed (or never fully opened) it will
+            # raise a AttributeError because it tries to call close()
+            # on self.socket, which is None
+            pass
```

Comment:
The only `self.socket = None` I see in my version of asyncore.dispatcher on Python 2.7 is in `__init__`!

Reply:
Which is what I think is going on here: One of the tests is set up to throw an exception, and it does so from someplace before the object has been connected, so the socket is still None. And when it tries to close it--bam!